File: js/ai/src/message.ts
/**
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { extractJson } from './extract';
import type {
  MessageData,
  Part,
  ToolRequestPart,
  ToolResponsePart,
} from './model';

export type MessageParser<T = unknown> = (message: Message) => T;

/**
 * Message represents a single role's contribution to a generation. Each message
 * can contain multiple parts (for example text and an image), and each generation
 * can contain multiple messages.
 */
export class Message<T = unknown> implements MessageData {
  role: MessageData['role'];
  content: Part[];
  metadata?: Record<string, any>;
  parser?: MessageParser<T>;

  static parseData(
    lenientMessage:
      | string
      | (MessageData & { content: string | Part | Part[]; role: string })
      | MessageData,
    defaultRole: MessageData['role'] = 'user'
  ): MessageData {
    if (typeof lenientMessage === 'string') {
      return { role: defaultRole, content: [{ text: lenientMessage }] };
    }
    return {
      ...lenientMessage,
      content: Message.parseContent(lenientMessage.content),
    };
  }

  static parse(
    lenientMessage: string | (MessageData & { content: string }) | MessageData
  ): Message {
    return new Message(Message.parseData(lenientMessage));
  }

  static parseContent(lenientPart: string | Part | (string | Part)[]): Part[] {
    if (typeof lenientPart === 'string') {
      return [{ text: lenientPart }];
    } else if (Array.isArray(lenientPart)) {
      return lenientPart.map((p) => (typeof p === 'string' ? { text: p } : p));
    } else {
      return [lenientPart];
    }
  }

  constructor(message: MessageData, options?: { parser?: MessageParser<T> }) {
    this.role = message.role;
    this.content = message.content;
    this.metadata = message.metadata;
    this.parser = options?.parser;
  }

  /**
   * Attempts to parse the content of the message according to the supplied
   * output parser. Without a parser, returns `data` contained in the message or
   * tries to parse JSON from the text of the message.
   *
   * @returns The structured output contained in the message.
   */
  get output(): T {
    return this.parser?.(this) || this.data || extractJson<T>(this.text);
  }

  toolResponseParts(): ToolResponsePart[] {
    const res = this.content.filter((part) => !!part.toolResponse);
    return res as ToolResponsePart[];
  }

  /**
   * Concatenates all `text` parts present in the message with no delimiter.
   * @returns A string of all concatenated text parts.
   */
  get text(): string {
    return this.content.map((part) => part.text || '').join('');
  }

  /**
   * Concatenates all `reasoning` parts present in the message with no delimiter.
   * @returns A string of all concatenated reasoning parts.
   */
  get reasoning(): string {
    return this.content.map((part) => part.reasoning || '').join('');
  }

  /**
   * Returns the first media part detected in the message. Useful for extracting
   * (for example) an image from a generation expected to create one.
   * @returns The first detected `media` part in the message.
   */
  get media(): { url: string; contentType?: string } | null {
    return this.content.find((part) => part.media)?.media || null;
  }

  /**
   * Returns the first detected `data` part of a message.
   * @returns The first `data` part detected in the message (if any).
   */
  get data(): T | null {
    return this.content.find((part) => part.data)?.data as T | null;
  }

  /**
   * Returns all tool request found in this message.
   * @returns Array of all tool request found in this message.
   */
  get toolRequests(): ToolRequestPart[] {
    return this.content.filter(
      (part) => !!part.toolRequest
    ) as ToolRequestPart[];
  }

  /**
   * Returns all tool requests annotated with interrupt metadata.
   * @returns Array of all interrupt tool requests.
   */
  get interrupts(): ToolRequestPart[] {
    return this.toolRequests.filter((t) => !!t.metadata?.interrupt);
  }

  /**
   * Converts the Message to a plain JS object.
   * @returns Plain JS object representing the data contained in the message.
   */
  toJSON(): MessageData {
    const out: MessageData = {
      role: this.role,
      content: [...this.content],
    };
    if (this.metadata) out.metadata = this.metadata;
    return out;
  }
}

File: java/ai/src/main/java/com/google/genkit/ai/Message.java
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.google.genkit.ai;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * Message represents a message in a conversation with a generative AI model.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public class Message {

  @JsonProperty("role")
  private Role role;

  @JsonProperty("content")
  private List<Part> content = new ArrayList<>();

  @JsonProperty("metadata")
  private Map<String, Object> metadata;

  /**
   * Default constructor.
   */
  public Message() {
  }

  /**
   * Creates a message with the given role and content.
   *
   * @param role
   *            the message role
   * @param content
   *            the content parts
   */
  public Message(Role role, List<Part> content) {
    this.role = role;
    this.content = content != null ? new ArrayList<>(content) : new ArrayList<>();
  }

  /**
   * Creates a user message with text content.
   *
   * @param text
   *            the text content
   * @return a new user message
   */
  public static Message user(String text) {
    return new Message(Role.USER, Collections.singletonList(Part.text(text)));
  }

  /**
   * Creates a system message with text content.
   *
   * @param text
   *            the text content
   * @return a new system message
   */
  public static Message system(String text) {
    return new Message(Role.SYSTEM, Collections.singletonList(Part.text(text)));
  }

  /**
   * Creates a model message with text content.
   *
   * @param text
   *            the text content
   * @return a new model message
   */
  public static Message model(String text) {
    return new Message(Role.MODEL, Collections.singletonList(Part.text(text)));
  }

  /**
   * Creates a tool message with content.
   *
   * @param content
   *            the content parts
   * @return a new tool message
   */
  public static Message tool(List<Part> content) {
    return new Message(Role.TOOL, content);
  }

  /**
   * Returns the text content from all text parts.
   *
   * @return the concatenated text
   */
  public String getText() {
    if (content == null || content.isEmpty()) {
      return "";
    }
    StringBuilder sb = new StringBuilder();
    for (Part part : content) {
      if (part.getText() != null) {
        sb.append(part.getText());
      }
    }
    return sb.toString();
  }

  // Getters and setters

  public Role getRole() {
    return role;
  }

  public void setRole(Role role) {
    this.role = role;
  }

  public List<Part> getContent() {
    return content;
  }

  public void setContent(List<Part> content) {
    this.content = content;
  }

  public Map<String, Object> getMetadata() {
    return metadata;
  }

  public void setMetadata(Map<String, Object> metadata) {
    this.metadata = metadata;
  }

  /**
   * Adds a part to the message content.
   *
   * @param part
   *            the part to add
   * @return this message for chaining
   */
  public Message addPart(Part part) {
    if (this.content == null) {
      this.content = new ArrayList<>();
    }
    this.content.add(part);
    return this;
  }

  /**
   * Creates a builder for Message.
   *
   * @return a new builder
   */
  public static Builder builder() {
    return new Builder();
  }

  /**
   * Builder for Message.
   */
  public static class Builder {
    private Role role;
    private List<Part> content = new ArrayList<>();
    private Map<String, Object> metadata;

    public Builder role(Role role) {
      this.role = role;
      return this;
    }

    public Builder content(List<Part> content) {
      this.content = new ArrayList<>(content);
      return this;
    }

    public Builder addPart(Part part) {
      this.content.add(part);
      return this;
    }

    public Builder addText(String text) {
      this.content.add(Part.text(text));
      return this;
    }

    public Builder metadata(Map<String, Object> metadata) {
      this.metadata = metadata;
      return this;
    }

    public Message build() {
      Message message = new Message(role, content);
      message.setMetadata(metadata);
      return message;
    }
  }
}
